File: Chip/app/src/main/java/com/robertlevonyan/views/chip/ChipUtils.java
package com.robertlevonyan.views.chip;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffColorFilter;
import android.graphics.PorterDuffXfermode;
import android.graphics.Rect;
import android.graphics.RectF;
import android.graphics.drawable.Drawable;
import android.widget.ImageView;

/**
 * Created by robert on 2/27/2017.
 */

public class ChipUtils {

    static final int IMAGE_ID = 0x00910518;
    static final int TEXT_ID = 0x00059118;

    private static int[] colors = {0xd32f2f, 0xC2185B, 0x7B1FA2, 0x512DA8, 0x303F9F, 0x1976D2, 0x0288D1, 0x0097A7, 0x00796B, 0x388E3C, 0x689F38,
            0xAFB42B, 0xFBC02D, 0xFFA000, 0xF57C00, 0xE64A19, 0x5D4037, 0x616161, 0x455A64};

    public static Bitmap getScaledBitmap(Context context, Bitmap bitmap) {
        int size = context.getResources().getDimensionPixelSize(R.dimen.chip_height);
        return Bitmap.createScaledBitmap(bitmap, size, size, false);
    }

    public static Bitmap getSquareBitmap(Bitmap bitmap) {
        final Bitmap output;

        if (bitmap.getWidth() >= bitmap.getHeight()) {
            output = Bitmap.createBitmap(
                    bitmap,
                    bitmap.getWidth() / 2 - bitmap.getHeight() / 2,
                    0,
                    bitmap.getHeight(),
                    bitmap.getHeight()
            );
        } else {
            output = Bitmap.createBitmap(
                    bitmap,
                    0,
                    bitmap.getHeight() / 2 - bitmap.getWidth() / 2,
                    bitmap.getWidth(),
                    bitmap.getWidth()
            );
        }
        return output;
    }

    public static Bitmap getCircleBitmap(Context context, Bitmap bitmap, float radius) {
        int size = context.getResources().getDimensionPixelSize(R.dimen.chip_height);
        final Bitmap output = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888);
        final Canvas canvas = new Canvas(output);

        final int color = Color.RED;
        final Paint paint = new Paint();
        final Rect rect = new Rect(0, 0, size, size);
        final RectF rectF = new RectF(rect);

        paint.setAntiAlias(true);
        canvas.drawARGB(0, 0, 0, 0);
        paint.setColor(color);
        canvas.drawRoundRect(rectF, radius, radius, paint);

        paint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_IN));
        canvas.drawBitmap(bitmap, rect, rect, paint);

        return output;
    }

    public static Bitmap getCircleBitmapWithText(Context context, String text, int textColor, int bgColor, float radius) {
        int size = context.getResources().getDimensionPixelSize(R.dimen.chip_height);
        final Bitmap output = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888);
        final Canvas canvas = new Canvas(output);

        final Paint paint = new Paint();
        final Paint textPaint = new Paint();
        final Rect rect = new Rect(0, 0, size, size);
        final RectF rectF = new RectF(rect);

        paint.setAntiAlias(true);
        canvas.drawARGB(0, 0, 0, 0);
        paint.setColor(bgColor);
        canvas.drawRoundRect(rectF, radius, radius, paint);
        textPaint.setColor(textColor);
        textPaint.setStrokeWidth(30);
        textPaint.setTextSize(45);
        paint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_OVER));
        textPaint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_ATOP));

        int xPos;
        int yPos;
        if (text.length() == 1) {
            xPos = (int) ((canvas.getWidth() / 2) + ((textPaint.descent() + textPaint.ascent()) / 2));
            yPos = (int) ((canvas.getHeight() / 2) - ((textPaint.descent() + textPaint.ascent()) / 2));
        } else {
            xPos = (int) ((canvas.getWidth() / 3) + ((textPaint.descent() + textPaint.ascent()) / 2));
            yPos = (int) ((canvas.getHeight() / 2) - ((textPaint.descent() + textPaint.ascent()) / 2));
        }

        canvas.drawBitmap(output, rect, rect, paint);
        canvas.drawText(text, xPos, yPos, textPaint);

        return output;
    }

    public static String generateText(String iconText) {
        if (iconText.isEmpty()) {
            throw new IllegalStateException("Icon text must have at least one symbol");
        }
        if (iconText.length() == 1 || iconText.length() == 2) {
            return iconText;
        }

        String[] parts = iconText.split(" ");
        if (parts.length == 1) {
            String text = parts[0];
            text = text.substring(0, 2);

            String f = text.substring(0, 1);
            String s = text.substring(1, 2);

            f = f.toUpperCase();
            s = s.toLowerCase();

            text = f.concat(s);

            return text;
        }
        String first = parts[0];
        String second = parts[1];

        first = first.substring(0, 1);
        first = first.toUpperCase();
        second = second.substring(0, 1);
        second = second.toUpperCase();

        return first.concat(second);

    }

    public static void setIconColor(ImageView icon, int color) {
        Drawable iconDrawable = icon.getDrawable();
        icon.setColorFilter(new PorterDuffColorFilter(color, PorterDuff.Mode.SRC_ATOP));
        icon.setImageDrawable(iconDrawable);
    }
}

File: Chip/src/main/java/com/robertlevonyan/views/chip/ChipUtils.kt
package com.robertlevonyan.views.chip

import android.graphics.*
import java.util.*

internal fun Bitmap.getScaledBitmap(size: Int): Bitmap = Bitmap.createScaledBitmap(this, size, size, false)

internal fun Bitmap.getSquareBitmap(): Bitmap = if (width >= height) {
  Bitmap.createBitmap(this, width / 2 - height / 2, 0, height, height)
} else {
  Bitmap.createBitmap(this, 0, height / 2 - width / 2, width, width)
}

internal fun Bitmap.getCircleBitmap(size: Int, radius: Float): Bitmap {
  val output = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888)
  val canvas = Canvas(output)

  val color = Color.RED
  val paint = Paint()
  val rect = Rect(0, 0, size, size)
  val rectF = RectF(rect)

  paint.isAntiAlias = true
  canvas.drawARGB(0, 0, 0, 0)
  paint.color = color
  canvas.drawRoundRect(rectF, radius, radius, paint)

  paint.xfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_IN)
  canvas.drawBitmap(this, rect, rect, paint)

  return output
}

internal fun getCircleBitmapWithText(size: Int, text: String, textColor: Int, bgColor: Int, radius: Float): Bitmap {
  val output = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888)
  val canvas = Canvas(output)

  val paint = Paint()
  val textPaint = Paint()
  val rect = Rect(0, 0, size, size)
  val rectF = RectF(rect)

  paint.isAntiAlias = true
  canvas.drawARGB(0, 0, 0, 0)
  paint.color = bgColor
  canvas.drawRoundRect(rectF, radius, radius, paint)
  textPaint.color = textColor
  textPaint.strokeWidth = 30f
  textPaint.textSize = 45f
  paint.xfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_OVER)
  textPaint.xfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_ATOP)

  val (xPos, yPos) =
    if (text.length == 1) {
      (canvas.width / 2) + ((textPaint.descent() + textPaint.ascent()) / 2) to
          (canvas.height / 2) - ((textPaint.descent() + textPaint.ascent()) / 2)
    } else {
      (canvas.width / 3) + ((textPaint.descent() + textPaint.ascent()) / 2) to
          (canvas.height / 2) - ((textPaint.descent() + textPaint.ascent()) / 2)
    }

  canvas.drawBitmap(output, rect, rect, paint)
  canvas.drawText(text, xPos, yPos, textPaint)

  return output
}

internal fun String.generateText(): String {
  if (this.isEmpty()) {
    throw IllegalStateException("Icon text must have at least one symbol")
  }
  if (this.length == 1 || this.length == 2) {
    return this
  }

  val parts = this.split(" ")
  if (parts.size == 1) {
    val text = parts[0].substring(0, 2)

    val f = text.substring(0, 1).uppercase(Locale.getDefault())
    val s = text.substring(1, 2).lowercase(Locale.getDefault())

    return "$f$s"
  }

  val first = parts[0].substring(0, 1).uppercase(Locale.getDefault())
  val second = parts[1].substring(0, 1).uppercase(Locale.getDefault())

  return "$first$second"
}
